### PR TITLE
Unified data API

### PR DESCRIPTION
**Pull request by bensku**
Unified data API and CustomDataHolder-based implementation. Please review if API is any good, before I proceed into integrating it into custom block and item systems.

### CODE REVIEW

File: src/common/datas/view.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      const type = typeof value;
+      if (type == 'string') {
+        holder.set(key, 'string', value);
+      } else if (type == 'number') {
+        holder.set(key, 'number', value);
+      } else if (type == 'boolean') {
+        holder.set(key, 'boolean', value);
+      } else {
+        holder.set(key, value.constructor as new () => any, value);
+      }
```
Comment (Ap3teus) [inline src/common/datas/view.ts:57]: This probably throws an error for `undefined` or `null`?

File: src/common/datas/view.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import { DataHolder } from './holder';
+
+/**
+ * Creates a typed view into a data holder. The view contains values that were
+ * in the data holder when this was called. New values assigned to the view are
+ * immediately reflected to the data holder. Mutations to values in view are
+ * not updated to the data holder; immutable values should be preferred.
```
Comment (Ap3teus) [inline src/common/datas/view.ts:7]: > Mutations to values in view are
not updated to the data holder; immutable values should be preferred.

While this is in theory good idea, my experience is that in a lot of real world scenarios the code will be a lot cleaner if we just allow mutating the objects directly. Compare
```ts
view.a = { ...view.a, b: { ...view.a.b, c: 1 } }
```
to
```ts
view.a.b.c = 1
```

Though obviously everything is fine as long as the highest level api (Blocks, CustomItems...) allows this kind of pattern

File: src/common/datas/holder.ts [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    switch (type) {
+      case 'boolean':
+        this.container.set(
+          containerKey,
+          PersistentDataType.INTEGER,
+          ((value ? 1 : 0) as unknown) as Integer,
+        );
+        break;
+      case 'integer':
+        this.container.set(containerKey, PersistentDataType.INTEGER, value);
```
Comment (Ap3teus) [inline src/common/datas/holder.ts:154]: This probably throws an error if the `value` is not actually an integer? If so, the "Java-exception" should probably be handled and the code should throw a "JS-exception", i.e. an `Error`-object. Or should it just round it an pretend nothing happened?

File: src/common/datas/view.ts (hunk 1 above)
Comment (bensku) [reply src/common/datas/view.ts:57]: I re-checked the code, and types don't permit undefined or null here. Will add separate delete function.

File: src/common/datas/view.ts (hunk 2 above)
Comment (bensku) [reply src/common/datas/view.ts:7]: After major implementation changes, this restriction has been lifted.

File: src/common/datas/holder.ts (hunk 3 above)
Comment (bensku) [reply src/common/datas/holder.ts:154]: I've added check to ensure it is actually an integer.

**Pull request merged**